### **Data Analyzing and Manipulating**

We will use the "House Sales in King County, USA" dataset from Kaggle.
<br>Kaggle dataset link:
<br>https://www.kaggle.com/datasets/harlfoxem/housesalesprediction

In [1]:
import pandas as pd
from IPython.display import display

In [2]:
df_houses = pd.read_csv("kc_house_data _rev.csv")

In [3]:
with pd.option_context('display.max_columns', None):
    display(df_houses.head())

id             date     price  bedrooms  bathrooms  sqft_living  \
0  7129300520  20141013T000000  221900.0         3       1.00       1180.0   
1  6414100192  20141209T000000  538000.0         3       2.25       2570.0   
2  5631500400  20150225T000000  180000.0         2       1.00        770.0   
3  2487200875  20141209T000000  604000.0         4       3.00       1960.0   
4  1954400510  20150218T000000  510000.0         3       2.00       1680.0   

   sqft_lot  floors  waterfront  view  condition  grade  sqft_above  \
0    5650.0     1.0         0.0   0.0        3.0      7      1180.0   
1    7242.0     2.0         0.0   0.0        3.0      7      2170.0   
2   10000.0     1.0         0.0   0.0        3.0      6       770.0   
3    5000.0     1.0         0.0   0.0        5.0      7      1050.0   
4    8080.0     1.0         0.0   0.0        3.0      8      1680.0   

   sqft_basement  yr_built  yr_renovated  zipcode      lat     long  \
0            0.0    1955.0             0    98178  47.5112 -122.257   
1          400.0    1951.0          1991    98125  47.7210 -122.319   
2            0.0    1933.0             0    98028  47.7379 -122.233   
3          910.0    1965.0             0    98136  47.5208 -122.393   
4            0.0    1987.0             0    98074  47.6168 -122.045   

   sqft_living15  sqft_lot15  
0           1340        5650  
1           1690        7639  
2           2720        8062  
3           1360        5000  
4           1800        7503

In [4]:
display(df_houses.info())

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 21613 entries, 0 to 21612
Data columns (total 21 columns):
 #   Column         Non-Null Count  Dtype  
---  ------         --------------  -----  
 0   id             21613 non-null  int64  
 1   date           21613 non-null  object 
 2   price          21605 non-null  float64
 3   bedrooms       21613 non-null  int64  
 4   bathrooms      21610 non-null  float64
 5   sqft_living    21607 non-null  float64
 6   sqft_lot       21608 non-null  float64
 7   floors         21610 non-null  float64
 8   waterfront     21609 non-null  float64
 9   view           21608 non-null  float64
 10  condition      21610 non-null  float64
 11  grade          21613 non-null  int64  
 12  sqft_above     21606 non-null  float64
 13  sqft_basement  21611 non-null  float64
 14  yr_built       21606 non-null  float64
 15  yr_renovated   21613 non-null  int64  
 16  zipcode        21613 non-null  int64  
 17  lat            21613 non-null  float64
 18  long  

None

*RangeIndex: 21613 entries, 0 to 21612*
<br>*Data columns (total 21 columns)*
<br>This tells us that there are 21613 rows and 21 columns

There are three columns in the info data:
<br>***Column*** is the data columns.
<br>***Non-Null Count*** is the non-null (missing) data count.
<br>***Dtype*** is the data type.

In [5]:
# To get the DataFrame shape:
df_shape = df_houses.shape
print(df_shape)

(21613, 21)


In [6]:
# The "values" object or instance returns a NumPy representation of the DataFrame values.
array_data = df_houses.values
print(array_data)

[[7129300520 '20141013T000000' 221900.0 ... -122.257 1340 5650]
 [6414100192 '20141209T000000' 538000.0 ... -122.319 1690 7639]
 [5631500400 '20150225T000000' 180000.0 ... -122.233 2720 8062]
 ...
 [1523300141 '20140623T000000' 402101.0 ... -122.299 1020 2007]
 [291310100 '20150116T000000' 400000.0 ... -122.069 1410 1287]
 [1523300157 '20141015T000000' 325000.0 ... -122.299 1020 1357]]


In [7]:
# It is a 2-D array with the following shape:
data_shape = array_data.shape
print(data_shape)

(21613, 21)


In [8]:
# To get the DataFrame column names, we use the "columns" object or instance:
df_columns_names = df_houses.columns
display(df_columns_names)

Index(['id', 'date', 'price', 'bedrooms', 'bathrooms', 'sqft_living',
       'sqft_lot', 'floors', 'waterfront', 'view', 'condition', 'grade',
       'sqft_above', 'sqft_basement', 'yr_built', 'yr_renovated', 'zipcode',
       'lat', 'long', 'sqft_living15', 'sqft_lot15'],
      dtype='object')

In [9]:
# To get the row numbers or names if they have names, we use the "index" object or instance:
df_rows_index = df_houses.index
display(df_rows_index)

RangeIndex(start=0, stop=21613, step=1)

In [10]:
diff_dtypes = pd.DataFrame({
    'ints': [3, None, 4],  # Becomes Float64 due to NaN unless specified as Int64
    'floats': [2.5, 9.1, None],
    'bools': [True, None, False],
    'dates': pd.to_datetime(["2024-02-12", None, "2025-02-01"]),
    'categories': pd.Categorical(["M", "F", "M"]),
    'strings': ["Tom", "Mary", None],
}, index=["data1", "data2", "data3"])

data_rows_names = diff_dtypes.index

display(diff_dtypes)
display(data_rows_names)

ints  floats  bools      dates categories strings
data1   3.0     2.5   True 2024-02-12          M     Tom
data2   NaN     9.1   None        NaT          F    Mary
data3   4.0     NaN  False 2025-02-01          M    None

Index(['data1', 'data2', 'data3'], dtype='object')

#### **To sort the DataFrame:**
We use the "sort_values" function by adding the column name that we want to sort the data according to:

In [ ]:
# It will sort the data in ascending order:
data_sorted_ascend = df_houses.sort_values("yr_built")
data_sorted_ascend.head(10)

id             date     price  bedrooms  bathrooms  \
16084  7519000225  20141030T000000  465000.0         3        1.0   
9005   5160700035  20150422T000000  431000.0         2        1.5   
14536  8806900040  20150406T000000  415000.0         3        2.0   
12764  6362900145  20150203T000000  450000.0         4        2.0   
3919   7883604065  20150501T000000  210000.0         2        1.0   
19385  2420069042  20150424T000000  240000.0         3        2.0   
13683  2024089011  20140826T000000  550000.0         5        1.0   
2961   1947300115  20140619T000000  464000.0         3        1.0   
3882   7129302095  20150213T000000  265000.0         3        1.0   
19598  7518507685  20150223T000000  400000.0         3        1.0   

       sqft_living  sqft_lot  floors  waterfront  view  ...  grade  \
16084       1580.0    3774.0     1.5         0.0   0.0  ...      6   
9005        1300.0    4000.0     1.5         0.0   0.0  ...      6   
14536       1410.0    4303.0     1.5         0.0   0.0  ...      7   
12764       1960.0    5008.0     1.0         0.0   0.0  ...      6   
3919        1100.0    6000.0     1.5         0.0   0.0  ...      6   
19385       1553.0    6550.0     1.0         0.0   0.0  ...      7   
13683       2150.0  262231.0     1.5         0.0   0.0  ...      7   
2961        1320.0    3625.0     2.0         0.0   0.0  ...      7   
3882        1122.0    6554.0     1.5         0.0   0.0  ...      5   
19598       1100.0    5100.0     2.0         0.0   0.0  ...      7   

       sqft_above  sqft_basement  yr_built  yr_renovated  zipcode      lat  \
16084      1580.0            0.0    1900.0             0    98117  47.6839   
9005       1300.0            0.0    1900.0             0    98144  47.5937   
14536      1410.0            0.0    1900.0             0    98108  47.5541   
12764       980.0          980.0    1900.0          1988    98144  47.5958   
3919       1100.0            0.0    1900.0             0    98108  47.5275   
19385      1553.0            0.0    1900.0          2001    98022  47.2056   
13683      2150.0            0.0    1900.0          2000    98065  47.5519   
2961       1320.0            0.0    1900.0             0    98122  47.6049   
3882       1122.0            0.0    1900.0             0    98118  47.5135   
19598      1100.0            0.0    1900.0             0    98117  47.6790   

          long  sqft_living15  sqft_lot15  
16084 -122.361           1580        3860  
9005  -122.301           1480        4000  
14536 -122.317           1660        4326  
12764 -122.299           1175        2315  
3919  -122.323           1280        6000  
19385 -121.994           1010       10546  
13683 -121.803           1460       46609  
2961  -122.288           1660        5438  
3882  -122.257           1610        5650  
19598 -122.386           1540        5100  

[10 rows x 21 columns]

In [12]:
# It will sort the data in descending order:
data_sorted_descend = df_houses.sort_values("yr_built", ascending=False)
data_sorted_descend.head(10)

id             date      price  bedrooms  bathrooms  \
14489  2770601530  20140826T000000   500000.0         2       2.25   
20369  4385700250  20150407T000000  1800000.0         4       3.50   
21432  7104100110  20150511T000000   899000.0         4       3.50   
20311  7132300525  20150411T000000   500000.0         3       1.75   
7526   9520900210  20141231T000000   614285.0         5       2.75   
19945  8011100047  20150306T000000   530000.0         4       2.75   
1763   1832100030  20140625T000000   597326.0         4       4.00   
14925  2770602360  20150421T000000   671000.0         4       2.75   
8425    558100090  20150312T000000   628000.0         5       2.75   
8039   1250200495  20140624T000000   455000.0         2       1.50   

       sqft_living  sqft_lot  floors  waterfront  view  ...  grade  \
14489       1570.0    1269.0     2.0         0.0   0.0  ...      9   
20369       3480.0    4000.0     2.0         0.0   0.0  ...      9   
21432       2490.0    5500.0     2.0         0.0   0.0  ...      9   
20311       1530.0     825.0     3.0         0.0   0.0  ...      8   
7526        2730.0    6401.0     2.0         0.0   0.0  ...      8   
19945       2740.0    7872.0     2.0         0.0   0.0  ...     10   
1763        3570.0    8250.0     2.0         0.0   0.0  ...     10   
14925       1890.0    1475.0     2.0         0.0   0.0  ...      9   
8425        2600.0    8160.0     2.0         0.0   0.0  ...      8   
8039        1200.0    1259.0     2.0         0.0   0.0  ...      8   

       sqft_above  sqft_basement  yr_built  yr_renovated  zipcode      lat  \
14489      1280.0          290.0    2015.0             0    98199  47.6514   
20369      2460.0         1020.0    2015.0             0    98112  47.6356   
21432      1780.0          710.0    2015.0             0    98136  47.5499   
20311      1530.0            0.0    2015.0             0    98144  47.5929   
7526       2730.0            0.0    2015.0             0    98072  47.7685   
19945      2740.0            0.0    2015.0             0    98056  47.4954   
1763       2860.0          710.0    2015.0             0    98040  47.5784   
14925      1200.0          690.0    2015.0             0    98199  47.6472   
8425       2600.0            0.0    2015.0             0    98133  47.7348   
8039       1000.0          200.0    2015.0             0    98144  47.6001   

          long  sqft_living15  sqft_lot15  
14489 -122.385           1570        6000  
20369 -122.281           2620        4000  
21432 -122.393           1710        5500  
20311 -122.308           1580        1915  
7526  -122.160           2520        6126  
19945 -122.172           1220        6300  
1763  -122.226           2230       10000  
14925 -122.383           1650        1682  
8425  -122.340           1600        8160  
8039  -122.298           1320        1852  

[10 rows x 21 columns]

**Sorting by multiple columns:**

We will use the "House Sales in King County, USA" dataset from Kaggle.
<br>Kaggle dataset link:
<br>https://www.kaggle.com/datasets/harlfoxem/housesalesprediction

In [13]:
import pandas as pd
from IPython.display import display

In [14]:
df_houses = pd.read_csv("kc_house_data _rev.csv")

In [15]:
data_sorted_mult = df_houses.sort_values(["bedrooms", "price"])
data_sorted_mult.head(30)

id             date      price  bedrooms  bathrooms  \
14423  9543000205  20150413T000000   139950.0         0       0.00   
19452  3980300371  20140926T000000   142000.0         0       0.00   
4868   6896300380  20141002T000000   228000.0         0       1.00   
9854   7849202190  20141223T000000   235000.0         0       0.00   
8484   2310060040  20140925T000000   240000.0         0       2.50   
18379  1222029077  20141029T000000   265000.0         0       0.75   
3467   1453602309  20140805T000000   288000.0         0       1.50   
12653  7849202299  20150218T000000   320000.0         0       2.50   
8477   2569500210  20141117T000000   339950.0         0       2.50   
9773   3374500520  20150429T000000   355000.0         0       0.00   
3119   3918400017  20150205T000000   380000.0         0       0.00   
875    6306400140  20140612T000000  1100000.0         0       0.00   
6994   2954400190  20140624T000000  1300000.0         0       0.00   
1149   3421079032  20150217T000000    75000.0         1       0.00   
465    8658300340  20140523T000000    80000.0         1       0.75   
3108   1721801591  20150219T000000    89950.0         1       1.00   
7992   2734100835  20150303T000000    90000.0         1       1.00   
16530  2114700500  20150418T000000    90000.0         1       1.00   
3805   7335400215  20150505T000000    95000.0         1       0.75   
11416  7211402105  20141126T000000   106000.0         1       1.00   
13057  2853600155  20140915T000000   110000.0         1       1.00   
10140  2648500030  20140725T000000   112000.0         1       1.00   
12208  4031000520  20140708T000000   115000.0         1       2.00   
8080   1545807810  20141021T000000   118000.0         1       1.00   
9820   3598600049  20141003T000000   124000.0         1       0.75   
4651    859000110  20141002T000000   125000.0         1       1.00   
11739  1352300990  20140826T000000   126000.0         1       1.00   
16952  2722059292  20140604T000000   129000.0         1       1.00   
4089    985001015  20140604T000000   135000.0         1       1.00   
1066   3340401535  20141105T000000   140000.0         1       1.00   

       sqft_living  sqft_lot  floors  waterfront  view  ...  grade  \
14423        844.0    4269.0     1.0         0.0   0.0  ...      7   
19452        290.0   20875.0     1.0         0.0   0.0  ...      1   
4868         390.0    5900.0     1.0         0.0   0.0  ...      4   
9854        1470.0    4800.0     2.0         0.0   0.0  ...      7   
8484        1810.0    5669.0     2.0         0.0   0.0  ...      7   
18379        384.0  213444.0     1.0         0.0   0.0  ...      4   
3467        1430.0    1650.0     3.0         0.0   0.0  ...      7   
12653       1490.0    7111.0     2.0         0.0   0.0  ...      7   
8477        2290.0    8319.0     2.0         0.0   0.0  ...      8   
9773        2460.0    8049.0     2.0         0.0   0.0  ...      8   
3119        1470.0     979.0     3.0         0.0   2.0  ...      8   
875         3064.0    4764.0     3.5         0.0   2.0  ...      7   
6994        4810.0   28008.0     2.0         0.0   0.0  ...     12   
1149         670.0   43377.0     1.0         0.0   0.0  ...      3   
465          430.0    5050.0     1.0         0.0   0.0  ...      4   
3108         570.0    4080.0     1.0         0.0   0.0  ...      5   
7992         780.0    4000.0     1.0         0.0   0.0  ...      5   
16530        560.0    4120.0     1.0         0.0   0.0  ...      4   
3805         760.0    5746.0     1.0         0.0   0.0  ...      5   
11416        560.0    5700.0     1.0         0.0   0.0  ...      5   
13057        640.0   10280.0     1.0         0.0   0.0  ...      5   
10140       1080.0    3230.0     1.0         0.0   0.0  ...      6   
12208       1150.0    9812.0     1.0         0.0   0.0  ...      7   
8080         670.0    7957.0     1.0         0.0   0.0  ...      6   
9820         840.0    7203.0     1.5         0.0   0.0  ...      6   
4651         500.0    74

In [16]:
# Sorting multiple columns in ascending and descending order:
data_sorted_asc_desc = df_houses.sort_values(["bedrooms", "price"], ascending=[True, False])
data_sorted_asc_desc.head(30)

id             date      price  bedrooms  bathrooms  \
6994   2954400190  20140624T000000  1300000.0         0       0.00   
875    6306400140  20140612T000000  1100000.0         0       0.00   
3119   3918400017  20150205T000000   380000.0         0       0.00   
9773   3374500520  20150429T000000   355000.0         0       0.00   
8477   2569500210  20141117T000000   339950.0         0       2.50   
12653  7849202299  20150218T000000   320000.0         0       2.50   
3467   1453602309  20140805T000000   288000.0         0       1.50   
18379  1222029077  20141029T000000   265000.0         0       0.75   
8484   2310060040  20140925T000000   240000.0         0       2.50   
9854   7849202190  20141223T000000   235000.0         0       0.00   
4868   6896300380  20141002T000000   228000.0         0       1.00   
19452  3980300371  20140926T000000   142000.0         0       0.00   
14423  9543000205  20150413T000000   139950.0         0       0.00   
8445    724069023  20150414T000000  1250000.0         1       1.25   
9888   1231000310  20140812T000000   713000.0         1       1.00   
4410   4083304190  20140804T000000   680000.0         1       2.50   
8165   1797500530  20150505T000000   655100.0         1       1.00   
19273  9164100035  20150429T000000   655000.0         1       1.00   
9563   3023039231  20140714T000000   650000.0         1       1.00   
16381  6329000050  20150310T000000   641500.0         1       1.00   
5168   1720800305  20141119T000000   611900.0         1       2.25   
8955     59000201  20150505T000000   611206.0         1       1.00   
16680  4045500715  20141217T000000   598800.0         1       1.00   
18645  7942600910  20141216T000000   575000.0         1       1.00   
18597  3222049087  20150422T000000   570000.0         1       1.00   
5752   7624700050  20141223T000000   565000.0         1       1.00   
13843  8122100355  20140924T000000   550000.0         1       1.00   
17804  5561401530  20141029T000000   550000.0         1       1.50   
14858  3626039268  20140516T000000   540000.0         1       1.00   
12311  9523102420  20140922T000000   535000.0         1       1.00   

       sqft_living  sqft_lot  floors  waterfront  view  ...  grade  \
6994        4810.0   28008.0     2.0         0.0   0.0  ...     12   
875         3064.0    4764.0     3.5         0.0   2.0  ...      7   
3119        1470.0     979.0     3.0         0.0   2.0  ...      8   
9773        2460.0    8049.0     2.0         0.0   0.0  ...      8   
8477        2290.0    8319.0     2.0         0.0   0.0  ...      8   
12653       1490.0    7111.0     2.0         0.0   0.0  ...      7   
3467        1430.0    1650.0     3.0         0.0   0.0  ...      7   
18379        384.0  213444.0     1.0         0.0   0.0  ...      4   
8484        1810.0    5669.0     2.0         0.0   0.0  ...      7   
9854        1470.0    4800.0     2.0         0.0   0.0  ...      7   
4868         390.0    5900.0     1.0         0.0   0.0  ...      4   
19452        290.0   20875.0     1.0         0.0   0.0  ...      1   
14423        844.0    4269.0     1.0         0.0   0.0  ...      7   
8445        1810.0    5070.0     1.5         1.0   4.0  ...      8   
9888        1180.0    4000.0     1.5         0.0   2.0  ...      8   
4410        1820.0    3008.0     2.0         0.0   0.0  ...      7   
8165        1220.0    4160.0     1.0         0.0   0.0  ...      7   
19273       1660.0    5422.0     1.0         0.0   0.0  ...      7   
9563         920.0   91476.0     1.5         0.0   0.0  ...      6   
16381       1000.0    9084.0     1.0         1.0   3.0  ...      7   
5168        1220.0    2100.0     2.0         0.0   2.0  ...      8   
8955        1940.0    6300.0     1.0         0.0   3.0  ...      8   
16680       1090.0   32010.0     1.0         0.0   0.0  ...      6   
18645       1310.0    8667.0     1.5         0.0   0.0  ...      6   
18597        720.0    7540.0     1.0         1.0   4.0  ...      6   
5752        1370.0    62

**Subsetting columns:**
<br>To get specific columns:

In [17]:
df_price = df_houses["price"]
df_price.head()

0    221900.0
1    538000.0
2    180000.0
3    604000.0
4    510000.0
Name: price, dtype: float64

In [18]:
# Multiple columns:
names_col = ["price", "yr_built", "date"]
df_price_year_date = df_houses[names_col]

# or

# df_price_year_date = df_houses[["price", "yr_built", "date"]]
df_price_year_date.head()

price  yr_built             date
0  221900.0    1955.0  20141013T000000
1  538000.0    1951.0  20141209T000000
2  180000.0    1933.0  20150225T000000
3  604000.0    1965.0  20141209T000000
4  510000.0    1987.0  20150218T000000

**Subsetting rows:**
<br>To get specific rows, you can use the comparison operators, (==, !=, >, >=, <, <=):

In [19]:
years_larger_than_2014 = df_houses[df_houses["yr_built"] > 2014]

# or

# data_boolean = df_houses["yr_built"] > 2014
# # years_larger_than_2014 = df_houses[data_boolean]

years_larger_than_2014

id             date      price  bedrooms  bathrooms  \
643    9385200045  20150512T000000   729500.0         3       2.50   
1763   1832100030  20140625T000000   597326.0         4       4.00   
2687   3076500830  20141029T000000   385195.0         1       1.00   
4154   8077100031  20150422T000000   631000.0         3       2.25   
7526   9520900210  20141231T000000   614285.0         5       2.75   
8039   1250200495  20140624T000000   455000.0         2       1.50   
8425    558100090  20150312T000000   628000.0         5       2.75   
14215  8156600210  20150326T000000  1290000.0         5       3.50   
14489  2770601530  20140826T000000   500000.0         2       2.25   
14925  2770602360  20150421T000000   671000.0         4       2.75   
17098  9126100346  20140617T000000   350000.0         3       2.00   
19722  9161100075  20150318T000000   673000.0         4       2.25   
19805  9126100765  20140801T000000   455000.0         3       1.75   
19863  2768100510  20150402T000000   649000.0         3       2.00   
19907  7853440140  20150409T000000   802945.0         5       3.50   
19945  8011100047  20150306T000000   530000.0         4       2.75   
19959  7853430690  20150127T000000   572800.0         3       2.50   
20086  2768100512  20150422T000000   659000.0         2       2.50   
20251  7853440050  20150505T000000   771005.0         5       4.50   
20311  7132300525  20150411T000000   500000.0         3       1.75   
20369  4385700250  20150407T000000  1800000.0         4       3.50   
20487  9103000455  20150424T000000   920000.0         4       3.25   
20488  8691440220  20150202T000000  1290000.0         5       4.00   
20698   255470030  20150429T000000   619990.0         4       2.75   
20770  9310300160  20140828T000000   357000.0         5       2.50   
20798   952005863  20150505T000000   643950.0         3       2.25   
20852  1257201420  20140709T000000   595000.0         4       3.25   
20881  3052700385  20150414T000000   765000.0         4       2.25   
20933  2652501565  20150423T000000  1550000.0         3       3.25   
20963  6058600220  20140731T000000   230000.0         3       1.50   
21262  5694500840  20141125T000000   559000.0         2       3.00   
21270  1773100541  20150417T000000   389950.0         3       2.25   
21316  6600000050  20150310T000000  1700000.0         4       3.50   
21333  2424059174  20150508T000000  2000000.0         4       3.25   
21369   923059259  20150401T000000   455950.0         4       2.50   
21372  6169901185  20140520T000000   490000.0         5       3.50   
21432  7104100110  20150511T000000   899000.0         4       3.50   
21501  4233600150  20150203T000000  1150000.0         5       4.25   

       sqft_living  sqft_lot  floors  waterfront  view  ...  grade  \
643         1660.0    1091.0     3.0         0.0   1.0  ...      9   
1763        3570.0    8250.0     2.0         0.0   0.0  ...     10   
2687         710.0    6000.0     1.5         0.0   0.0  ...      6   
4154        1670.0    1396.0     2.0         0.0   0.0  ...      9   
7526        2730.0    6401.0     2.0         0.0   0.0  ...      8   
8039        1200.0    1259.0     2.0         0.0   0.0  ...      8   
8425        2600.0    8160.0     2.0         0.0   0.0  ...      8   
14215       2980.0    5100.0     2.0         0.0   0.0  ...     10   
14489       1570.0    1269.0     2.0         0.0   0.0  ...      9   
14925       1890.0    1475.0     2.0         0.0   0.0  ...      9   
17098       1380.0    3600.0     3.0         0.0   0.0  ...      8   
19722       2580.0    2875.0     2.0         0.0   0.0  ...      9   
19805       1320.0    1014.0     3.0         0.0   0.0  ...      9   
19863       1530.0    1442.0     3.0         0.0   0.0  ...      9   
19907       4000.0    9234.0     2.0         0.0   0.0  ...      9   
19945       2740.0    7872.0     2.0         0.0   0.0  ...     10   
19959       3310.0    4682.0     2.0         0.0   0.0  ...      9   
20086       1450.0    12

We can use Bitwise Operators, (& (and), | (or), ~ (not) as follows:

In [20]:
years_equal_2014 = df_houses["yr_built"] == 2014
prices_equal_to = df_houses["price"] == 861990

df_price_year = df_houses[years_equal_2014 & prices_equal_to]
df_price_year

id             date     price  bedrooms  bathrooms  sqft_living  \
42  7203220400  20140707T000000  861990.0         5       2.75       3595.0   

    sqft_lot  floors  waterfront  view  ...  grade  sqft_above  sqft_basement  \
42    5639.0     2.0         0.0   0.0  ...      9      3595.0            0.0   

    yr_built  yr_renovated  zipcode      lat     long  sqft_living15  \
42    2014.0             0    98053  47.6848 -122.016           3625   

    sqft_lot15  
42        5639  

[1 rows x 21 columns]

**Subsetting using the "isin" method:**
<br>To get a specific rows:

In [21]:
bedrooms_num_bool = df_houses["bedrooms"].isin([5, 2])
bedrooms_num_5_2 = df_houses[bedrooms_num_bool]

# This will return all the house that has 5 or 2 bedrooms:
bedrooms_num_5_2

id             date      price  bedrooms  bathrooms  \
2      5631500400  20150225T000000   180000.0         2       1.00   
11     9212900260  20140527T000000   468000.0         2       1.00   
14     1175000570  20150312T000000   530000.0         5       2.00   
18       16000397  20141205T000000   189000.0         2       1.00   
22     7137970340  20140703T000000   285000.0         5       2.50   
...           ...              ...        ...       ...        ...   
21595  1972201967  20141031T000000   520000.0         2       2.25   
21596  7502800100  20140813T000000   679950.0         5       2.75   
21600   249000205  20141015T000000  1540000.0         5       3.75   
21610  1523300141  20140623T000000   402101.0         2       0.75   
21612  1523300157  20141015T000000   325000.0         2       0.75   

       sqft_living  sqft_lot  floors  waterfront  view  ...  grade  \
2            770.0   10000.0     1.0         0.0   0.0  ...      6   
11          1160.0    6000.0     1.0         0.0   0.0  ...      7   
14          1810.0    4850.0     1.5         0.0   0.0  ...      7   
18          1200.0    9850.0     1.0         0.0   0.0  ...      7   
22          2270.0    6300.0     2.0         0.0   0.0  ...      8   
...            ...       ...     ...         ...   ...  ...    ...   
21595       1530.0     981.0     3.0         0.0   0.0  ...      8   
21596       3600.0    9437.0     2.0         0.0   0.0  ...      9   
21600       4470.0    8088.0     2.0         0.0   0.0  ...     11   
21610       1020.0    1350.0     2.0         0.0   0.0  ...      7   
21612       1020.0    1076.0     2.0         0.0   0.0  ...      7   

       sqft_above  sqft_basement  yr_built  yr_renovated  zipcode      lat  \
2           770.0            0.0    1933.0             0    98028  47.7379   
11          860.0          300.0    1942.0             0    98115  47.6900   
14         1810.0            0.0    1900.0             0    98107  47.6700   
18         1200.0            0.0    1921.0             0    98002  47.3089   
22         2270.0            0.0       NaN             0    98092  47.3266   
...           ...            ...       ...           ...      ...      ...   
21595      1480.0           50.0    2006.0             0    98103  47.6533   
21596      3600.0            0.0    2014.0             0    98059  47.4822   
21600      4470.0            0.0    2008.0             0    98004  47.6321   
21610      1020.0            0.0    2009.0             0    98144  47.5944   
21612      1020.0            0.0    2008.0             0    98144  47.5941   

          long  sqft_living15  sqft_lot15  
2     -122.233           2720        8062  
11    -122.292           1330        6000  
14    -122.394           1360        4850  
18    -122.210           1060        5095  
22    -122.169           2240        7005  
...        ...            ...         ...  
21595 -122.346           1530        1282  
21596 -122.131           3550        9421  
21600 -122.200           2780        8964  
21610 -122.299           1020        2007  
21612 -122.299           1020        1357  

[4361 rows x 21 columns]

**Adding new columns:**

In [22]:
df_houses["total_area"] = df_houses["sqft_living"] + df_houses["sqft_lot"] + df_houses["sqft_above"] + df_houses["sqft_basement"]

df_houses.columns

Index(['id', 'date', 'price', 'bedrooms', 'bathrooms', 'sqft_living',
       'sqft_lot', 'floors', 'waterfront', 'view', 'condition', 'grade',
       'sqft_above', 'sqft_basement', 'yr_built', 'yr_renovated', 'zipcode',
       'lat', 'long', 'sqft_living15', 'sqft_lot15', 'total_area'],
      dtype='object')

**Get unique values:**
<br>The "unique" function will return the unique values in a column.

In [23]:
unique_years = df_houses["yr_built"].unique()
unique_years

array([1955., 1951., 1933., 1965., 1987., 2001., 1995., 1963., 1960.,
         nan, 1942., 1927., 1977., 1900., 1979., 1994., 1916., 1921.,
       1969., 1947., 1968., 1985., 1941., 1915., 1909., 1948., 2005.,
       2003., 1929., 1981., 1930., 1904., 1996., 2000., 2014., 1922.,
       1959., 1966., 1953., 1950., 2008., 1991., 1954., 1973., 1925.,
       1989., 1972., 1986., 1956., 2002., 1992., 1964., 1952., 1961.,
       2006., 1988., 1962., 1939., 1946., 1967., 1975., 1980., 1910.,
       1983., 1984., 1905., 1971., 2010., 1945., 1924., 1990., 1914.,
       1978., 1926., 2004., 1923., 2007., 1949., 1999., 1901., 1993.,
       1920., 1997., 1943., 1976., 1957., 1940., 1918., 1928., 1974.,
       1911., 1936., 1937., 1982., 1908., 1931., 1998., 1913., 2013.,
       1907., 1958., 2012., 1912., 2011., 1917., 1932., 1944., 1902.,
       2009., 1903., 1970., 2015., 1934., 1938., 1919., 1906., 1935.])